sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1612,9 +1612,12 @@ object ConvertToLocalRelation extends Rule[LogicalPlan] {
  * {{{
  *   SELECT DISTINCT f1, f2 FROM t  ==>  SELECT f1, f2 FROM t GROUP BY f1, f2
  * }}}
+ *
+ * The [[Distinct]] can be ignored if it is the right child of a left semi or anti join.
  */
 object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
-  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case j @ Join(_, Distinct(right), LeftSemiOrAnti(_), _, _) => j.copy(right = right)
```

Comment:
Just out of curiosity, this rule can always make performance better? For example, how about the case where a right table has too many duplicates?

Reply:
I benchmarked a bit and you are right - this can cause a performance regression. For example q87 took a 10% hit on sf100. I think it's best to close this PR and issue.